File: backend/modules/progress/src/main/java/com/kineticos/progress/domain/ProgressEntry.java
package com.kineticos.progress.domain;

import lombok.Getter;

import java.math.BigDecimal;
import java.time.Instant;

/**
 * Métrica de seguimiento con histórico (tabla progress_entries): peso, IMC, % grasa,
 * circunferencias, etc.
 */
@Getter
public class ProgressEntry {

    private final Long id;
    private final Long userId;
    private final String metricType;
    private final BigDecimal value;
    private final String unit;
    private final Instant measuredAt;
    private final String notes;
    private final Instant createdAt;

    private ProgressEntry(Long id, Long userId, String metricType, BigDecimal value, String unit,
                          Instant measuredAt, String notes, Instant createdAt) {
        this.id = id;
        this.userId = userId;
        this.metricType = metricType;
        this.value = value;
        this.unit = unit;
        this.measuredAt = measuredAt;
        this.notes = notes;
        this.createdAt = createdAt;
    }

    public static ProgressEntry create(Long userId, String metricType, BigDecimal value, String unit,
                                       Instant measuredAt, String notes) {
        return new ProgressEntry(null, userId, metricType, value, unit,
                measuredAt == null ? Instant.now() : measuredAt, notes, Instant.now());
    }

    public static ProgressEntry restore(Long id, Long userId, String metricType, BigDecimal value, String unit,
                                        Instant measuredAt, String notes, Instant createdAt) {
        return new ProgressEntry(id, userId, metricType, value, unit, measuredAt, notes, createdAt);
    }

    public boolean belongsTo(Long userId) {
        return this.userId.equals(userId);
    }
}

File: mobile/lib/features/progress/domain/progress_entry.dart
// =====================================================================
// KineticOs — Métrica de seguimiento (mirror de ProgressDtos en backend).
// =====================================================================
class ProgressEntry {
  final int id;
  final String metricType;
  final double value;
  final String unit;
  final DateTime measuredAt;
  final String? notes;

  const ProgressEntry({
    required this.id,
    required this.metricType,
    required this.value,
    required this.unit,
    required this.measuredAt,
    this.notes,
  });

  factory ProgressEntry.fromJson(Map<String, dynamic> json) => ProgressEntry(
        id: json['id'] as int,
        metricType: json['metricType'] as String,
        value: (json['value'] as num).toDouble(),
        unit: json['unit'] as String,
        measuredAt: DateTime.parse(json['measuredAt'] as String),
        notes: json['notes'] as String?,
      );
}
